Convert to GitHub Actions

marcguyer opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation |no
| Bugfix        |no
| BC Break      |no
| New Feature   |no
| RFC           |no
| QA            | yes

### Description

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE master BRANCH

- Are you adding documentation?
  - TARGET THE master BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE master BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE master BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

Add GHA CI workflow according to https://gist.github.com/weierophinney/b003e50c3c2667d08076caf31ebd36a4

marcguyer committed:
Convert to GitHub Actions

Signed-off-by: Marc Guyer <marcguyer@gmail.com>

froschdesign commented:
@marcguyer 
The readme file must also be updated:

```markdown
[![Build Status](https://github.com/mezzio/mezzio-authentication-oauth2/workflows/Continuous%20Integration/badge.svg)](https://github.com/mezzio/mezzio-authentication-oauth2/actions?query=workflow%3A"Continuous+Integration")
```
…and remove the old badges.

Thanks in advance! 👍

boesing commented on composer.json:
```diff
@@ -1,3 +1,5 @@
+
+
```
You might want to revert these empty-line changes

boesing commented on .laminas-ci.json:
```diff
@@ -0,0 +1,5 @@
+{
+    "extensions": [
+         "pdo-sqlite",
```
Maybe GHA is fails due to the trailing `,` but not 100% sure about it, might be only the empty lines in `composer.json` tho.

boesing commented:
Hm, lowest dependencies worked back in january:

https://travis-ci.com/github/mezzio/mezzio-authentication-oauth2/jobs/472985387

Same as latest:

https://travis-ci.com/github/mezzio/mezzio-authentication-oauth2/jobs/472985388


Maybe comparing `composer show` output of travis and GHA might help.

marcguyer commented:
The run for php8 will need to use `lcobucci/jwt` v4 or better: https://github.com/mezzio/mezzio-authentication-oauth2/pull/28/checks?check_run_id=2254965808#step:3:303

The latest Travis build used 4.0.0 correctly: https://travis-ci.com/github/mezzio/mezzio-authentication-oauth2/jobs/472985387#L223

The failing GA build incorrectly uses 3.4.0.

So far, I'm not sure what, if anything, I can do to fix that. Seems like a composer issue. Matrix looks good to me but I'm just looking at this for the first time.

boesing commented:
Well, thats a bummer.

# This is what happens
 
We are supporting two versions of `lcobucci/jwt` to fulfill the constraints of this package.
1. v3 (supports PHP 7.3)
2. v4 (supports PHP 7.4 & PHP 8)

Since we are using `composer update --ignore-platform-req=php --prefer-lowest` to install the dependencies for PHP 8.0, the 3.x version of `lcobucci/jwt` is being installed, even tho that it is not compatible with PHP 8.0.

We still have to use `--ignore-platform-req=php` as some 3rd-party (or even laminas packages) are not yet PHP 8.0 compatible.

That might be the same issue with the `locked` dependencies, as it contains the `3.x` version of `lcobucci/jwt` aswell.
The fact, that `lcobucci/jwt` dropped PHP 7.3 support in `4.0` is a problem as this package supports PHP 7.3 and thus has to use the 3.x version within the `composer.lock` file.

There is no way of having proper locked dependencies for this package as there is no version of `lcobucci/jwt` which fulfills all constraints.

----

@weierophinney @Ocramius what is the best way to fix this? Yes, we could remove support for PHP 7.3 but that would violate what we somewhere decided to do: "avoid dropping support for officially supported PHP versions by php.net" (or something like that).

marcguyer committed:
Exclude failing tests

Lowest and locked deps on the php8 runs are impossible due to dependency conflicts.

Signed-off-by: Marc Guyer <marcguyer@gmail.com>

Danielss89 commented:
What does "dropping support" mean?
Is it security fixes only, or is it bug fixes too?
If it is only security fixes, i would just create a new branch, for the current major version, and do security fixes in this branch going forward.
Then bump the current major and use this for future development and bugfixes.

weierophinney commented:
@boesing This is an interesting case because this package itself does not directly depend on lcobucci/jwt; it's an indirect dependency required by league/oauth2-server.

In this particular case, if you run `composer update --prefer-lowest` with  PHP 8.0, one of two things will happen:

- It will fail, because some packages do not support PHP 8 yet.
- It will work, **and** you'll get the v4 release of lcobucci/jwt installed due to the fact that it's the lowest version that PHP 8 supports.

I just tried it locally, and it actually works, and, as I noted, I get v4 of the jwt library.

As such, I think this patch is fine. Eventually, we'll need to no longer run `composer update` with the `--ignore-platform-reqs` flag in the CI action; I think I'm going to go ahead and update the ci-matrix and continuous-integration actions to allow setting a config flag to disable usage of that option, so we can test packages like this that where installation will work without it.

weierophinney committed:
qa: do not ignore platform reqs on PHP 8

New features on laminas-ci now allow this.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: update dependencies to pick up lcobucci/jwt 4.x versions

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: exclude locked run on PHP 7.3

So that we get an lcobucci/jwt version that works with this PHP version.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
Alright, here's what I've done:

- laminas-ci-matrix-action and laminas-continuous-integration-action now support a new config flag, "ignore_platform_reqs_8". When not present, it has the existing behavior (pass the `--ignore-platform-req=php` option); when present, the value of the flag indicates whether or not that option is passed.
- I've added `"ignore_platform_reqs_8": false` to the `.laminas-ci.json`.
- I've updated the lockfile to pickup the lcobucci/jwt 4.x version.
- I now only exclude the locked run on PHP 7.3.

Tests pass, and we run against more PHP version/dependency combinations, giving greater confidence that changes going forward will work correctly.
